pkg/envoy/lds/builder.go

```diff
@@ -0,0 +1,594 @@
+package lds
+
+import (
+	"errors"
+	"fmt"
+
+	xds_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	xds_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	xds_http_local_ratelimit "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/local_ratelimit/v3"
+	xds_hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	xds_tcp_proxy "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/tcp_proxy/v3"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	configv1alpha2 "github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	policyv1alpha1 "github.com/openservicemesh/osm/pkg/apis/policy/v1alpha1"
+	"github.com/openservicemesh/osm/pkg/auth"
+	"github.com/openservicemesh/osm/pkg/identity"
+	"github.com/openservicemesh/osm/pkg/service"
+
+	"github.com/openservicemesh/osm/pkg/envoy"
+	"github.com/openservicemesh/osm/pkg/protobuf"
+	"github.com/openservicemesh/osm/pkg/trafficpolicy"
+)
+
+const (
+	websocketUpgradeType = "websocket"
+)
+
+func ListenerBuilder() *listenerBuilder { //nolint: revive // unexported-return
```

Comment:
I am envisioning the builder to be instantiated from outside the pkg at some point, most likely after we introduce the snapshot cache mechanism. So the function is exported. the type isn't exported because the setters in the builder may perform additional logic than just setting the field (e.g. AddFilters in HttpConnManagerBuilder). It's irrelevant at this moment because both the caller of the builder and the builder live in the same pkg.

Reply:
Right, I'm saying that that reasoning may be useful in a comment